frame/dvm/evm/src/lib.rs

```diff
@@ -0,0 +1,656 @@
+		let account_id = T::AddressMapping::into_account_id(*address);
+
+		let nonce = frame_system::Module::<T>::account_nonce(&account_id);
+		let balance = T::Currency::free_balance(&account_id);
+
+		Account {
+			nonce: U256::from(UniqueSaturatedInto::<u128>::unique_saturated_into(nonce)),
+			balance: U256::from(UniqueSaturatedInto::<u128>::unique_saturated_into(balance)),
+		}
+	}
+
+	fn mutate_account_basic(address: &H160, new: Account) {
+		let account_id = T::AddressMapping::into_account_id(*address);
+		let current = T::AccountBasicMapping::account_basic(address);
+
+		if current.nonce < new.nonce {
+			// ASSUME: in one single EVM transaction, the nonce will not increase more than
+			// `u128::max_value()`.
+			for _ in 0..(new.nonce - current.nonce).low_u128() {
+				frame_system::Module::<T>::inc_account_nonce(&account_id);
+			}
+		}
+
+		if current.balance > new.balance {
+			let diff = current.balance - new.balance;
+			T::Currency::slash(&account_id, diff.low_u128().unique_saturated_into());
+		} else if current.balance < new.balance {
+			let diff = new.balance - current.balance;
+			T::Currency::deposit_creating(&account_id, diff.low_u128().unique_saturated_into());
+		}
+	}
+}
+
+/// Substrate system chain ID.
+pub struct SystemChainId;
+
+impl Get<u64> for SystemChainId {
+	fn get() -> u64 {
+		sp_io::misc::chain_id()
+	}
+}
+
+static ISTANBUL_CONFIG: Config = Config::istanbul();
+
+/// EVM module trait
+pub trait Trait: frame_system::Trait + pallet_timestamp::Trait {
+	/// Calculator for current gas price.
+	type FeeCalculator: FeeCalculator;
+
+	/// Allow the origin to call on behalf of given address.
+	type CallOrigin: EnsureAddressOrigin<Self::Origin>;
+	/// Allow the origin to withdraw on behalf of given address.
+	type WithdrawOrigin: EnsureAddressOrigin<Self::Origin, Success = Self::AccountId>;
+
+	/// Mapping from address to account id.
+	type AddressMapping: AddressMapping<Self::AccountId>;
+	/// Currency type for withdraw and balance storage.
+	type Currency: Currency<Self::AccountId>;
+
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+	/// Precompiles associated with this EVM engine.
+	type Precompiles: Precompiles;
+	/// Chain ID of EVM.
+	type ChainId: Get<u64>;
+	/// The account basic mapping way
+	type AccountBasicMapping: AccountBasicMapping;
+
+	/// EVM config used in the module.
+	fn config() -> &'static Config {
+		&ISTANBUL_CONFIG
+	}
+}
+
+#[cfg(feature = "std")]
+#[derive(Clone, Eq, PartialEq, Encode, Decode, Debug, Serialize, Deserialize)]
+/// Account definition used for genesis block construction.
+pub struct GenesisAccount {
+	/// Account nonce.
+	pub nonce: U256,
+	/// Account balance.
+	pub balance: U256,
+	/// Full account storage.
+	pub storage: std::collections::BTreeMap<H256, H256>,
+	/// Account code.
+	pub code: Vec<u8>,
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as EVM {
```

Comment:
Should consider using a different storage name to identify from `evm` pallet.

Reply:
Done.